# Missing Data

## 0.0 Load libaries and connect your google drive

In [15]:
%matplotlib inline


import matplotlib.pyplot as plt
import pandas as pd
import numpy as np

import sklearn as skl
import scipy as sci
import scipy.io as sio

#configure the plotting defaluts
plt.rcParams['figure.figsize'] = [15.256 , 6.024]
plt.rcParams['figure.dpi'] = 125 

In [16]:
from google.colab import drive

drive.mount('/content/drive/', force_remount=True)

data_path='/content/drive/My Drive/2022_SFC_IDSP/datasets/'
results_path='/content/drive/My Drive/2022_SFC_IDSP/Week_2/'

Mounted at /content/drive/


### 0.1 About the data

1.   It is good practice to always figure out a few things about a data set that you are working with. 
2.   This data comes the the [US Fuel Economy website](https://www.fueleconomy.gov/feg/download.shtml). This provides feul economy (and emissions) data for vechicles sold in the US since 1984. So in some sense this is a 'big data'$^1$ set in the sense that it is not easily understood or searchable with out some data tools. 
3. you should have found this data set on the moodle page as well as a text file explaining all the columns of the data. (vechiles.csv and Us_gov_fuel_economy_data_description.docx)  

It makes sense to now load and have a peak at it and then examine the metadata file to understand what you are looking at. (This split of data description and raw data should make sense based on the [Data Carpentry reading from Week 1](https://datacarpentry.org/spreadsheet-ecology-lesson/01-format-data/index.html))

You can either take a peek at the data in a spreadsheet tool (like sheets in google), or what I would recommend is load it here (your first action of this lecture).

---


Footnote:

$^1$ **big data** is usually means gigabites of information, but for the purposes of this course **big** or **large** data will be used to refer to data sets which are too big to easily be viewed in a single screen. From a cognative point of view understanding the trends in the are difficult to process.

#### 0.5 Example load the data
Load the 'vehicles.csv' dataset using the pandas command ```.read_csv```. 

In [31]:
vechile_mpg = pd.read_csv(data_path + 'vehicles.csv' )

vechile_mpg

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2882: DtypeWarning: Columns (70,71,72,73,74,76,79) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


barrels08  barrelsA08  charge120  charge240  city08  city08U  cityA08  \
0      14.167143         0.0        0.0        0.0      19      0.0        0   
1      27.046364         0.0        0.0        0.0       9      0.0        0   
2      11.018889         0.0        0.0        0.0      23      0.0        0   
3      27.046364         0.0        0.0        0.0      10      0.0        0   
4      15.658421         0.0        0.0        0.0      17      0.0        0   
...          ...         ...        ...        ...     ...      ...      ...   
44782  13.523182         0.0        0.0        0.0      19      0.0        0   
44783  12.935217         0.0        0.0        0.0      20      0.0        0   
44784  14.167143         0.0        0.0        0.0      18      0.0        0   
44785  14.167143         0.0        0.0        0.0      18      0.0        0   
44786  16.528333         0.0        0.0        0.0      16      0.0        0   

       cityA08U  cityCD  cityE  ...  mfrCode  c240Dscr  charge240b  c240bDscr  \
0           0.0     0.0    0.0  ...      NaN       NaN         0.0        NaN   
1           0.0     0.0    0.0  ...      NaN       NaN         0.0        NaN   
2           0.0     0.0    0.0  ...      NaN       NaN         0.0        NaN   
3           0.0     0.0    0.0  ...      NaN       NaN         0.0        NaN   
4           0.0     0.0    0.0  ...      NaN       NaN         0.0        NaN   
...         ...     ...    ...  ...      ...       ...         ...        ...   
44782       0.0     0.0    0.0  ...      NaN       NaN         0.0        NaN   
44783       0.0     0.0    0.0  ...      NaN       NaN         0.0        NaN   
44784       0.0     0.0    0.0  ...      NaN       NaN         0.0        NaN   
44785       0.0     0.0    0.0  ...      NaN       NaN         0.0        NaN   
44786       0.0     0.0    0.0  ...      NaN       NaN         0.0        NaN   

                          createdOn                    modifiedOn  startStop  \
0      Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013        NaN   
1      Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013        NaN   
2      Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013        NaN   
3      Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013        NaN   
4      Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013        NaN   
...                             ...                           ...        ...   
44782  Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013        NaN   
44783  Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013        NaN   
44784  Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013        NaN   
44785  Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013        NaN   
44786  Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013        NaN   

       phevCity  phevHwy  phevComb  
0             0        0         0  
1             0        0         0  
2             0        0         0  
3             0        0         0  
4             0        0         0  
...         ...      ...       ...  
44782         0        0         0  
44783         0        0         0  
44784         0        0         0  
44785         0        0         0  
44786         0        0         0  

[44787 rows x 83 columns]

In [ ]:
vechile_mpg.columns

Index(['barrels08', 'barrelsA08', 'charge120', 'charge240', 'city08',
       'city08U', 'cityA08', 'cityA08U', 'cityCD', 'cityE', 'cityUF', 'co2',
       'co2A', 'co2TailpipeAGpm', 'co2TailpipeGpm', 'comb08', 'comb08U',
       'combA08', 'combA08U', 'combE', 'combinedCD', 'combinedUF', 'cylinders',
       'displ', 'drive', 'engId', 'eng_dscr', 'feScore', 'fuelCost08',
       'fuelCostA08', 'fuelType', 'fuelType1', 'ghgScore', 'ghgScoreA',
       'highway08', 'highway08U', 'highwayA08', 'highwayA08U', 'highwayCD',
       'highwayE', 'highwayUF', 'hlv', 'hpv', 'id', 'lv2', 'lv4', 'make',
       'model', 'mpgData', 'phevBlended', 'pv2', 'pv4', 'range', 'rangeCity',
       'rangeCityA', 'rangeHwy', 'rangeHwyA', 'trany', 'UCity', 'UCityA',
       'UHighway', 'UHighwayA', 'VClass', 'year', 'youSaveSpend', 'guzzler',
       'trans_dscr', 'tCharger', 'sCharger', 'atvType', 'fuelType2', 'rangeA',
       'evMotor', 'mfrCode', 'c240Dscr', 'charge240b', 'c240bDscr',
       'createdOn', 'modifiedOn

In [6]:
full_data_summary =vechile_mpg.describe()
full_data_summary

barrels08    barrelsA08  charge120     charge240        city08  \
count  44787.000000  44787.000000    44787.0  44787.000000  44787.000000   
mean      15.400525      0.191517        0.0      0.089215     18.876973   
std        4.281554      0.981003        0.0      0.877820      9.506672   
min        0.047081      0.000000        0.0      0.000000      6.000000   
25%       12.935217      0.000000        0.0      0.000000     15.000000   
50%       14.875500      0.000000        0.0      0.000000     17.000000   
75%       17.500588      0.000000        0.0      0.000000     21.000000   
max       42.501429     16.528333        0.0     15.300000    150.000000   

            city08U       cityA08      cityA08U        cityCD         cityE  \
count  44787.000000  44787.000000  44787.000000  44787.000000  44787.000000   
mean       7.497398      0.836761      0.704402      0.000509      0.591830   
std       13.438829      6.408916      6.317163      0.037682      5.240771   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.000000      0.000000      0.000000      0.000000      0.000000   
75%       16.479600      0.000000      0.000000      0.000000      0.000000   
max      150.195800    145.000000    145.083500      5.350000    122.000000   

       ...         UCity        UCityA      UHighway     UHighwayA  \
count  ...  44787.000000  44787.000000  44787.000000  44787.000000   
mean   ...     23.933411      1.110191     35.045874      0.889549   
std    ...     13.519488      9.056496     12.716474      5.661646   
min    ...      0.000000      0.000000      0.000000      0.000000   
25%    ...     18.500000      0.000000     28.000000      0.000000   
50%    ...     21.790100      0.000000     33.500000      0.000000   
75%    ...     26.198100      0.000000     39.600000      0.000000   
max    ...    224.800000    207.262200    187.100000    173.143600   

               year  youSaveSpend    charge240b      phevCity       phevHwy  \
count  44787.000000  44787.000000  44787.000000  44787.000000  44787.000000   
mean    2003.128966  -5675.017304      0.016378      0.266394      0.268917   
std       11.947750   4888.737166      0.330668      3.748301      3.627405   
min     1984.000000 -41500.000000      0.000000      0.000000      0.000000   
25%     1992.000000  -8250.000000      0.000000      0.000000      0.000000   
50%     2004.000000  -5250.000000      0.000000      0.000000      0.000000   
75%     2014.000000  -2500.000000      0.000000      0.000000      0.000000   
max     2023.000000   9750.000000      9.600000     97.000000     81.000000   

          phevComb  
count  44787.00000  
mean       0.26635  
std        3.67006  
min        0.00000  
25%        0.00000  
50%        0.00000  
75%        0.00000  
max       88.00000  

[8 rows x 59 columns]

In [7]:
full_data_summary.loc['count']<44787

barrels08          False
barrelsA08         False
charge120          False
charge240          False
city08             False
city08U            False
cityA08            False
cityA08U           False
cityCD             False
cityE              False
cityUF             False
co2                False
co2A               False
co2TailpipeAGpm    False
co2TailpipeGpm     False
comb08             False
comb08U            False
combA08            False
combA08U           False
combE              False
combinedCD         False
combinedUF         False
cylinders           True
displ               True
engId              False
feScore            False
fuelCost08         False
fuelCostA08        False
ghgScore           False
ghgScoreA          False
highway08          False
highway08U         False
highwayA08         False
highwayA08U        False
highwayCD          False
highwayE           False
highwayUF          False
hlv                False
hpv                False
id                 False


## 1.0 Selecting the specific columns of the data

As you can see the dataframe is of the shape (44787, 83). If you display the dataframe you will notice that jupyter even has some difficulty in presenting this many columns. 

 Inital inspection of several of these columns shows that they appear all 0 or blank. Reading the metadata, shows that this has to do with the orignal table being adjust to account for electric vechicles and changing fuel standards.

For this part of the lecture we really only need to examine a couple of coulumns to offer some ideas of how to work with 'missing' data. To do that we will create a new dataframe by copying the columns that are of interest. I am going to select the following colums:



*   'year'
*   'make'
*   'model'
*   'city08' (city mpg)
*   'highway08' (highway mpg)
*   'fuelType'
*   'cylinders'
*   'displ'



---
We are going to use the ``` .copy()``` function to create a subset of the 'vechile_mpg' dataframe. Everything before the ``` .copy()``` command is creating a **view** of the data.  This is not an actual new bit of memory, rather we are telling the computer to look at a specific sub-section of the memory allocated the the eintire 'vichile_mpg' dataframe. From this point forward we want to modify and change the new dataframe that we are creating, not the orignal dataset. To do that we need to **copy** a selection of 'vichile_mpg' and place it in a new bit of memory. This highlights the object oriented nature of Python; we can have two objects looking at the same bit of memory. So if we change one, it is possiable to change the other. To avoid this we need to allocate more memory.

A good discussion of this can be found here:

https://realpython.com/pandas-settingwithcopywarning/




####  1.0.1 Example selecting a subset of the  dataset and creating a new dataframe 

In [35]:
simple_mpg= vechile_mpg[['year','make', 'model', 'city08','highway08', 'fuelType', 'cylinders','displ']].copy()

simple_mpg

year        make                model  city08  highway08 fuelType  \
0      1985  Alfa Romeo   Spider Veloce 2000      19         25  Regular   
1      1985     Ferrari           Testarossa       9         14  Regular   
2      1985       Dodge              Charger      23         33  Regular   
3      1985       Dodge  B150/B250 Wagon 2WD      10         12  Regular   
4      1993      Subaru     Legacy AWD Turbo      17         23  Premium   
...     ...         ...                  ...     ...        ...      ...   
44782  1993      Subaru               Legacy      19         26  Regular   
44783  1993      Subaru               Legacy      20         28  Regular   
44784  1993      Subaru           Legacy AWD      18         24  Regular   
44785  1993      Subaru           Legacy AWD      18         24  Regular   
44786  1993      Subaru     Legacy AWD Turbo      16         21  Premium   

       cylinders  displ  
0            4.0    2.0  
1           12.0    4.9  
2            4.0    2.2  
3            8.0    5.2  
4            4.0    2.2  
...          ...    ...  
44782        4.0    2.2  
44783        4.0    2.2  
44784        4.0    2.2  
44785        4.0    2.2  
44786        4.0    2.2  

[44787 rows x 8 columns]

#### 1.0.2 Example using describe to explore the data

Here we use the describe function to get a sense what the data is doing from a summary statistics point of view.

In [36]:
simple_mpg.describe()

year        city08     highway08     cylinders         displ
count  44787.000000  44787.000000  44787.000000  44415.000000  44417.000000
mean    2003.128966     18.876973     24.958158      5.710999      3.283853
std       11.947750      9.506672      8.818173      1.768306      1.357762
min     1984.000000      6.000000      9.000000      2.000000      0.000000
25%     1992.000000     15.000000     20.000000      4.000000      2.200000
50%     2004.000000     17.000000     24.000000      6.000000      3.000000
75%     2014.000000     21.000000     28.000000      6.000000      4.200000
max     2023.000000    150.000000    133.000000     16.000000      8.400000

**observe** that since there are still 44787 rows it is hard to explore the data table visually.  Even using the magic wand to convert that dataframe into an interactive data table (ie something even more similar to a spread sheet. So we need a way to figure out if there is any data missing in our dataset.

####  1.0.3 Action Save off the new **dataset**

####  1.0.4 Action Load the new **dataset** and inspect the resutling dataframe

In [ ]:
simple_mpg

In [ ]:
simple_mpg.describe()

## 2.0 Finding the missing data

**Observe** that both the 'cylinders'  and 'displ' (i.e. engine displacement) series, only have a count of 44415 elements, not the 44787 rows that the full data set.  This simple tool has told us that for the data we have possesses missing values, but **where** and **Why**

To answer that lets introduce a new function ```.isna()```. This will search through the data and tell use where the missing values are. For the rest of the example I am going to focus on the 'cylinders' series. You will need to repeat for the 'displ' series.

#### 2.0.1 Example of the ``` .isna()``` function

In [6]:
simple_mpg['cylinders'].isna()

0        False
1        False
2        False
3        False
4        False
         ...  
44782    False
44783    False
44784    False
44785    False
44786    False
Name: cylinders, Length: 44787, dtype: bool

We can see that this function produces a boolean series. But due to the length (44787 rows). we will have no clue as to where these missing values are. slicing out specific rows where the boolean value is 'True' should illuminate some things. 




#### 2.0.2 Example Use the ``` .loc ``` function to examine the rows missing data.


In [37]:
simple_mpg.loc[simple_mpg['cylinders'].isna()]

year    make                          model  city08  highway08  \
7138   2000  Nissan                       Altra EV      81         91   
7139   2000  Toyota                        RAV4 EV      81         64   
8143   2001  Toyota                        RAV4 EV      81         64   
8144   2001    Ford                          Th!nk      74         58   
8146   2001    Ford         Explorer USPS Electric      45         33   
...     ...     ...                            ...     ...        ...   
38616  2022    Ford              Mustang Mach-E GT      90         77   
38617  2022    Ford  Mustang Mach-E GT Performance      88         75   
38618  2022    Ford             Mustang Mach-E RWD     110         96   
38619  2022    Ford    Mustang Mach-E RWD Extended     104         90   
38675  2022   Tesla                    Model Y AWD     129        116   

          fuelType  cylinders  displ  
7138   Electricity        NaN    NaN  
7139   Electricity        NaN    NaN  
8143   Electricity        NaN    NaN  
8144   Electricity        NaN    NaN  
8146   Electricity        NaN    NaN  
...            ...        ...    ...  
38616  Electricity        NaN    NaN  
38617  Electricity        NaN    NaN  
38618  Electricity        NaN    NaN  
38619  Electricity        NaN    NaN  
38675  Electricity        NaN    NaN  

[372 rows x 8 columns]

### 2.1 Replacing missing values

Now can observe that the the missing values in the 'cylinders' column is due to these being electric cars.

At this point there are several options for how you can choose to deal with the missing data.  This decision will depend on the anlysis that you wish to persue, and the type of data that you have.  The possiable options are



1.   **Remove** all the rows with missing data using the ``.dropna()`` function.
2.    **Fill in**  missing values with specific value using ``.fillna()``
3.   **Interpolate** the missing values based on the rest of the data with ``.interpolate()``

All of these are valid options depending on the goals of your analysis.  Here we are just going to replace the NaN values with a fixed value of 0 since, electric vechiles have 0 cylinders.


#### 2.1.1 Example of filling in values

In [38]:
simple_mpg['cylinders'].fillna(0)

0         4.0
1        12.0
2         4.0
3         8.0
4         4.0
         ... 
44782     4.0
44783     4.0
44784     4.0
44785     4.0
44786     4.0
Name: cylinders, Length: 44787, dtype: float64

Now re-examining the data 

In [16]:
simple_mpg.loc[simple_mpg['cylinders'].isna() ]

Unnamed: 0  year    make                          model  city08  \
7138         7138  2000  Nissan                       Altra EV      81   
7139         7139  2000  Toyota                        RAV4 EV      81   
8143         8143  2001  Toyota                        RAV4 EV      81   
8144         8144  2001    Ford                          Th!nk      74   
8146         8146  2001    Ford         Explorer USPS Electric      45   
...           ...   ...     ...                            ...     ...   
38616       38616  2022    Ford              Mustang Mach-E GT      90   
38617       38617  2022    Ford  Mustang Mach-E GT Performance      88   
38618       38618  2022    Ford             Mustang Mach-E RWD     110   
38619       38619  2022    Ford    Mustang Mach-E RWD Extended     104   
38675       38675  2022   Tesla                    Model Y AWD     129   

       highway08     fuelType  cylinders  
7138          91  Electricity        NaN  
7139          64  Electricity        NaN  
8143          64  Electricity        NaN  
8144          58  Electricity        NaN  
8146          33  Electricity        NaN  
...          ...          ...        ...  
38616         77  Electricity        NaN  
38617         75  Electricity        NaN  
38618         96  Electricity        NaN  
38619         90  Electricity        NaN  
38675        116  Electricity        NaN  

[372 rows x 8 columns]

What happened? the function ran with out error, but the orignal dataframe is unchanged. In this case, it should be pointed out that many of the native pandas functions do not overwirte or **mutate** the orignal object. Hence, when you apply the function you end up creating a new bit of memory. (In this case it is a new, but unnamed series.)

To address this we use the **keyword parameter** 'inplace' and set it equal to True (note capital T).

#### 2.1.2 Example of using inplace to chage orignal dataframe column

In [39]:
simple_mpg['cylinders'].fillna(0,inplace=True)

In [40]:
simple_mpg.loc[simple_mpg['cylinders'].isna()]

Empty DataFrame
Columns: [year, make, model, city08, highway08, fuelType, cylinders, displ]
Index: []

Again the filtering above now prodices an empty dataframe, which makes sense as we have filled in the cylinders coulmn.  But we have not affected the 'displ' column.  That is left for you to deal with in the next set of actions.  But first lets resave our data frame.


#### 2.1.3 Example of historgram of the data

In [62]:
simple_mpg['cylinders']].plot.hist()

SyntaxError: ignored

#### 2.1.4 Action save modified data frame.

#### 2.1.5 Action - Fill in the missing values in the 'displ' column and save resulting data.





#### 2.1.6 Action explore the data using different plots. Consdier examing scatter plots comparing two different columns. Save off the most informative plots. 

#### 2.1.7 Action Go to this mentimeter site to discuss what information can be gain by further analysis and propose how you would do this in Excel (or prefered tools).

https://www.menti.com/t7tq4yfia4

